File: extension/src/shared/mainAgents.ts
import rawMainAgentCatalog from "../../../config/main-agents.json";

export type MainAgent = string;

export interface MainAgentConfig {
  id: MainAgent;
  label: string;
  description: string;
  remoteAliases: string[];
}

interface MainAgentCatalogFile {
  defaultMainAgent: string;
  mainAgents: Array<{
    id: string;
    label?: string;
    description?: string;
    remoteAliases?: string[];
  }>;
}

function normalizeMainAgentAlias(value: string) {
  return value.trim().toLowerCase().replace(/-/gu, "_");
}

function parseMainAgentCatalog(value: unknown): { defaultMainAgent: MainAgent; mainAgents: MainAgentConfig[] } {
  if (!value || typeof value !== "object") {
    throw new Error("Main agent catalog must be an object");
  }

  const { defaultMainAgent, mainAgents } = value as Partial<MainAgentCatalogFile>;
  if (typeof defaultMainAgent !== "string" || !defaultMainAgent.trim()) {
    throw new Error("Main agent catalog must declare a non-empty defaultMainAgent");
  }

  if (!Array.isArray(mainAgents) || !mainAgents.length) {
    throw new Error("Main agent catalog must declare at least one main agent");
  }

  const seenIds = new Set<string>();
  const parsedMainAgents = mainAgents.map((entry, index) => {
    if (!entry || typeof entry !== "object") {
      throw new Error(`Main agent entry at index ${index} must be an object`);
    }

    const id = entry.id?.trim();
    if (!id) {
      throw new Error(`Main agent entry at index ${index} must declare a non-empty id`);
    }
    if (seenIds.has(id)) {
      throw new Error(`Duplicate main agent id: ${id}`);
    }
    seenIds.add(id);

    const remoteAliases = Array.from(new Set([id, ...(entry.remoteAliases ?? [])]
      .map((alias) => typeof alias === "string" ? alias.trim() : "")
      .filter(Boolean)));
    if (!remoteAliases.length) {
      throw new Error(`Main agent ${id} must declare at least one remote alias`);
    }

    return Object.freeze({
      id,
      label: entry.label?.trim() || id,
      description: entry.description?.trim() || (id === defaultMainAgent ? "默认主 AGENT" : "可切换的备用主 AGENT"),
      remoteAliases
    });
  });

  if (!seenIds.has(defaultMainAgent)) {
    throw new Error(`defaultMainAgent ${defaultMainAgent} is not present in mainAgents`);
  }

  return {
    defaultMainAgent,
    mainAgents: parsedMainAgents
  };
}

const parsedMainAgentCatalog = parseMainAgentCatalog(rawMainAgentCatalog);

export const MAIN_AGENT_CONFIGS = Object.freeze(parsedMainAgentCatalog.mainAgents);
export const MAIN_AGENTS = Object.freeze(MAIN_AGENT_CONFIGS.map((agent) => agent.id));
export const DEFAULT_MAIN_AGENT: MainAgent = parsedMainAgentCatalog.defaultMainAgent;

const MAIN_AGENT_SET = new Set(MAIN_AGENTS);
const MAIN_AGENT_CONFIG_MAP = new Map(MAIN_AGENT_CONFIGS.map((agent) => [agent.id, agent]));
const MAIN_AGENT_ALIAS_MAP = new Map(
  MAIN_AGENT_CONFIGS.flatMap((agent) => agent.remoteAliases.map((alias) => [normalizeMainAgentAlias(alias), agent.id] as const))
);

export function isMainAgent(value: unknown): value is MainAgent {
  return typeof value === "string" && MAIN_AGENT_SET.has(value);
}

export function getMainAgentConfig(agent: string) {
  return MAIN_AGENT_CONFIG_MAP.get(agent) ?? null;
}

export function resolveConfiguredMainAgentByAlias(alias: string) {
  return MAIN_AGENT_ALIAS_MAP.get(normalizeMainAgentAlias(alias)) ?? null;
}

File: python_adapter/app/main_agents.py
from __future__ import annotations

import json
from pathlib import Path

from .config import ADAPTER_ROOT


MainAgent = str
MAIN_AGENT_CONFIG_PATH = ADAPTER_ROOT.parent / "config" / "main-agents.json"


def _normalize_main_agent_alias(value: str) -> str:
    return value.strip().lower().replace("-", "_")


def _load_main_agent_catalog(config_path: Path) -> tuple[MainAgent, tuple[dict[str, object], ...]]:
    payload = json.loads(config_path.read_text(encoding="utf-8"))
    if not isinstance(payload, dict):
        raise RuntimeError("Main agent catalog must be an object")

    default_main_agent = payload.get("defaultMainAgent")
    main_agents = payload.get("mainAgents")

    if not isinstance(default_main_agent, str) or not default_main_agent.strip():
        raise RuntimeError("Main agent catalog must declare a non-empty defaultMainAgent")
    if not isinstance(main_agents, list) or not main_agents:
        raise RuntimeError("Main agent catalog must declare at least one main agent")

    seen_ids: set[str] = set()
    parsed_main_agents: list[dict[str, object]] = []
    for index, entry in enumerate(main_agents):
        if not isinstance(entry, dict):
            raise RuntimeError(f"Main agent entry at index {index} must be an object")

        agent_id = entry.get("id")
        if not isinstance(agent_id, str) or not agent_id.strip():
            raise RuntimeError(f"Main agent entry at index {index} must declare a non-empty id")
        agent_id = agent_id.strip()
        if agent_id in seen_ids:
            raise RuntimeError(f"Duplicate main agent id: {agent_id}")
        seen_ids.add(agent_id)

        remote_aliases = entry.get("remoteAliases")
        aliases: list[str] = [agent_id]
        if isinstance(remote_aliases, list):
            aliases.extend(alias for alias in remote_aliases if isinstance(alias, str) and alias.strip())
        aliases = list(dict.fromkeys(alias.strip() for alias in aliases if alias.strip()))
        if not aliases:
            raise RuntimeError(f"Main agent {agent_id} must declare at least one remote alias")

        parsed_main_agents.append(
            {
                "id": agent_id,
                "label": entry.get("label", agent_id),
                "description": entry.get(
                    "description",
                    "默认主 AGENT" if agent_id == default_main_agent else "可切换的备用主 AGENT",
                ),
                "remoteAliases": aliases,
            }
        )

    if default_main_agent not in seen_ids:
        raise RuntimeError(f"defaultMainAgent {default_main_agent} is not present in mainAgents")

    return default_main_agent, tuple(parsed_main_agents)


DEFAULT_MAIN_AGENT, MAIN_AGENT_CONFIGS = _load_main_agent_catalog(MAIN_AGENT_CONFIG_PATH)
ALLOWED_MAIN_AGENTS = frozenset(str(agent["id"]) for agent in MAIN_AGENT_CONFIGS)
REMOTE_AGENT_WHITELIST: dict[MainAgent, frozenset[str]] = {
    str(agent["id"]): frozenset(_normalize_main_agent_alias(alias) for alias in list(agent["remoteAliases"]))
    for agent in MAIN_AGENT_CONFIGS
}